README.md
![Unisync logo](https://raw.githubusercontent.com/hyper-prog/unisync/master/doc/docimg/icon.png)

UniSync
=======

 UniSync is an open-source File synchronization 
 program for Linux and Windows. It is an open source utility 
 for efficiently comparing or synchronizing large
 directory structures by checking names and sizes or optionally
 the times or even the contents by hashes. (Md5 or Sha-2)
 The UniSync can even synchronize offline directories 
 (Which are not available same time) by creating a catalog file
 and making and update package according to that.

 Documentation: http://hyperprog.com/unisync/unisyncdoc/doc/unisync/using.html

 The UniSync has a graphical user interface names UniSyncGui.
 https://github.com/hyper-prog/unisyncgui
 
 ![UnisyncGui](https://raw.githubusercontent.com/hyper-prog/unisync/master/doc/docimg/unisync_gui.jpg)
 

Webpage: http://hyperprog.com/unisync

Compile / Install
-----------------

 Simply rename Makefile.linux to Makefile and hit "make" to compile it.

Author
------
 Péter Deák (http://hyperprog.com) hyper80@gmail.com

License
-------
 GPLv2

ChangeLog
1.0
    -Moved to github

0.971
    -Added -wl windows locale switch to display corrent character on buggy windows terminals

0.97
    -Added -i Interactive/paranoid sync switch.
     Prints a small statistics and ask user before do real sync.	

0.961
    -Fix open's third parameter bug

0.96
    -Dramatical copy/scan speed improvements using platfom dependent api.
    -Add -std switch to force standard POSIX routines
    -Fix the root directory scanning on windows/linux
     (The drive root was claimed to not existed directory)
    -File/Dir modification time fixes
    -Error handling fixes

0.953
    -Fixed stat error handling
    -Code style fixes

0.952
    -Added some more error messages when stat error occured.
    -Fixed missing program return code.

0.951
    -Typo fixes

0.95
    -Added -fixtime switch
    -Fix bug, near unsuccesful directory creating.

0.94
    -First release


md5.c
/*
  * This is an OpenSSL-compatible implementation of the RSA Data Security,
  * Inc. MD5 Message-Digest Algorithm.
  *
  * Written by Solar Designer <solar@openwall.com> in 2001, and placed in
  * the public domain.  See md5.c for more information.
  */

 /* Any 32-bit or wider unsigned integer data type will do */
 typedef unsigned long MD5_u32plus;

 typedef struct {
         MD5_u32plus lo, hi;
         MD5_u32plus a, b, c, d;
         unsigned char buffer[64];
         MD5_u32plus block[16];
 } MD5_CTX;

 /*
  * The basic MD5 functions.
  *
  * F is optimized compared to its RFC 1321 definition just like in Colin
  * Plumb's implementation.
  */
 #define F(x, y, z)                      ((z) ^ ((x) & ((y) ^ (z))))
 #define G(x, y, z)                      ((y) ^ ((z) & ((x) ^ (y))))
 #define H(x, y, z)                      ((x) ^ (y) ^ (z))
 #define I(x, y, z)                      ((y) ^ ((x) | ~(z)))

 /*
  * The MD5 transformation for all four rounds.
  */
 #define STEP(f, a, b, c, d, x, t, s) \
         (a) += f((b), (c), (d)) + (x) + (t); \
         (a) = (((a) << (s)) | (((a) & 0xffffffff) >> (32 - (s)))); \
         (a) += (b);

 /*
  * SET reads 4 input bytes in little-endian byte order and stores them
  * in a properly aligned word in host byte order.
  *
  * The check for little-endian architectures which tolerate unaligned
  * memory accesses is just an optimization.  Nothing will break if it
  * doesn't work.
  */
 #if defined(__i386__) || defined(__vax__)
 #define SET(n) \
         (*(MD5_u32plus *)&ptr[(n) * 4])
 #define GET(n) \
         SET(n)
 #else
 #define SET(n) \
         (ctx->block[(n)] = \
         (MD5_u32plus)ptr[(n) * 4] | \
         ((MD5_u32plus)ptr[(n) * 4 + 1] << 8) | \
         ((MD5_u32plus)ptr[(n) * 4 + 2] << 16) | \
         ((MD5_u32plus)ptr[(n) * 4 + 3] << 24))
 #define GET(n) \
         (ctx->block[(n)])
 #endif

 /*
  * This processes one or more 64-byte data blocks, but does NOT update
  * the bit counters.  There're no alignment requirements.
  */
 static void *body(MD5_CTX *ctx, void *data, unsigned long size)
 {
         unsigned char *ptr;
         MD5_u32plus a, b, c, d;
         MD5_u32plus saved_a, saved_b, saved_c, saved_d;

         ptr = (unsigned char *)data;

         a = ctx->a;
         b = ctx->b;
         c = ctx->c;
         d = ctx->d;

         do {
                 saved_a = a;
                 saved_b = b;
                 saved_c = c;
                 saved_d = d;

 /* Round 1 */
                 STEP(F, a, b, c, d, SET(0), 0xd76aa478, 7)
                 STEP(F, d, a, b, c, SET(1), 0xe8c7b756, 12)
                 STEP(F, c, d, a, b, SET(2), 0x242070db, 17)
                 STEP(F, b, c, d, a, SET(3), 0xc1bdceee, 22)
                 STEP(F, a, b, c, d, SET(4), 0xf57c0faf, 7)
                 STEP(F, d, a, b, c, SET(5), 0x4787c62a, 12)
                 STEP(F, c, d, a, b, SET(6), 0xa8304613, 17)
                 STEP(F, b, c, d, a, SET(7), 0xfd469501, 22)
                 STEP(F, a, b, c, d, SET(8), 0x698098d8, 7)
                 STEP(F, d, a, b, c, SET(9), 0x8b44f7af, 12)
                 STEP(F, c, d, a, b, SET(10), 0xffff5bb1, 17)
                 STEP(F, b, c, d, a, SET(11), 0x895cd7be, 22)
                 STEP(F, a, b, c, d, SET(12), 0x6b901122, 7)
                 STEP(F, d, a, b, c, SET(13), 0xfd987193, 12)
                 STEP(F, c, d, a, b, SET(14), 0xa679438e, 17)
                 STEP(F, b, c, d, a, SET(15), 0x49b40821, 22)

 /* Round 2 */
                 STEP(G, a, b, c, d, GET(1), 0xf61e2562, 5)
                 STEP(G, d, a, b, c, GET(6), 0xc040b340, 9)
                 STEP(G, c, d, a, b, GET(11), 0x265e5a51, 14)
                 STEP(G, b, c, d, a, GET(0), 0xe9b6c7aa, 20)
                 STEP(G, a, b, c, d, GET(5), 0xd62f105d, 5)
                 STEP(G, d, a, b, c, GET(10), 0x02441453, 9)
                 STEP(G, c, d, a, b, GET(15), 0xd8a1e681, 14)
                 STEP(G, b, c, d, a, GET(4), 0xe7d3fbc8, 20)
                 STEP(G, a, b, c, d, GET(9), 0x21e1cde6, 5)
                 STEP(G, d, a, b, c, GET(14), 0xc33707d6, 9)
                 STEP(G, c, d, a, b, GET(3), 0xf4d50d87, 14)
                 STEP(G, b, c, d, a, GET(8), 0x455a14ed, 20)
                 STEP(G, a, b, c, d, GET(13), 0xa9e3e905, 5)
                 STEP(G, d, a, b, c, GET(2), 0xfcefa3f8, 9)
                 STEP(G, c, d, a, b, GET(7), 0x676f02d9, 14)
                 STEP(G, b, c, d, a, GET(12), 0x8d2a4c8a, 20)

 /* Round 3 */
                 STEP(H, a, b, c, d, GET(5), 0xfffa3942, 4)
                 STEP(H, d, a, b, c, GET(8), 0x8771f681, 11)
                 STEP(H, c, d, a, b, GET(11), 0x6d9d6122, 16)
                 STEP(H, b, c, d, a, GET(14), 0xfde5380c, 23)
                 STEP(H, a, b, c, d, GET(1), 0xa4beea44, 4)
                 STEP(H, d, a, b, c, GET(4), 0x4bdecfa9, 11)
                 STEP(H, c, d, a, b, GET(7), 0xf6bb4b60, 16)
                 STEP(H, b, c, d, a, GET(10), 0xbebfbc70, 23)
                 STEP(H, a, b, c, d, GET(13), 0x289b7ec6, 4)
                 STEP(H, d, a, b, c, GET(0), 0xeaa127fa, 11)
                 STEP(H, c, d, a, b, GET(3), 0xd4ef3085, 16)
                 STEP(H, b, c, d, a, GET(6), 0x04881d05, 23)
                 STEP(H, a, b, c, d, GET(9), 0xd9d4d039, 4)
                 STEP(H, d, a, b, c, GET(12), 0xe6db99e5, 11)
                 STEP(H, c, d, a, b, GET(15), 0x1fa27cf8, 16)
                 STEP(H, b, c, d, a, GET(2), 0xc4ac5665, 23)

 /* Round 4 */
                 STEP(I, a, b, c, d, GET(0), 0xf4292244, 6)
                 STEP(I, d, a, b, c, GET(7), 0x432aff97, 10)
                 STEP(I, c, d, a, b, GET(14), 0xab9423a7, 15)
                 STEP(I, b, c, d, a, GET(5), 0xfc93a039, 21)
                 STEP(I, a, b, c, d, GET(12), 0x655b59c3, 6)
                 STEP(I, d, a, b, c, GET(3), 0x8f0ccc92, 10)
                 STEP(I, c, d, a, b, GET(10), 0xffeff47d, 15)
                 STEP(I, b, c, d, a, GET(1), 0x85845dd1, 21)
                 STEP(I, a, b, c, d, GET(8), 0x6fa87e4f, 6)
                 STEP(I, d, a, b, c, GET(15), 0xfe2ce6e0, 10)
                 STEP(I, c, d, a, b, GET(6), 0xa3014314, 15)
                 STEP(I, b, c, d, a, GET(13), 0x4e0811a1, 21)
                 STEP(I, a, b, c, d, GET(4), 0xf7537e82, 6)
                 STEP(I, d, a, b, c, GET(11), 0xbd3af235, 10)
                 STEP(I, c, d, a, b, GET(2), 0x2ad7d2bb, 15)
                 STEP(I, b, c, d, a, GET(9), 0xeb86d391, 21)

                 a += saved_a;
                 b += saved_b;
                 c += saved_c;
                 d += saved_d;

                 ptr += 64;
         } while (size -= 64);

         ctx->a = a;
         ctx->b = b;
         ctx->c = c;
         ctx->d = d;

         return ptr;
 }

 void MD5_Init(MD5_CTX *ctx)
 {
         ctx->a = 0x67452301;
         ctx->b = 0xefcdab89;
         ctx->c = 0x98badcfe;
         ctx->d = 0x10325476;

         ctx->lo = 0;
         ctx->hi = 0;
 }

 void MD5_Update(MD5_CTX *ctx, void *data, unsigned long size)
 {
         MD5_u32plus saved_lo;
         unsigned long used, free;

         saved_lo = ctx->lo;
         if ((ctx->lo = (saved_lo + size) & 0x1fffffff) < saved_lo)
                 ctx->hi++;
         ctx->hi += size >> 29;

         used = saved_lo & 0x3f;

         if (used) {
                 free = 64 - used;

                 if (size < free) {
                         memcpy(&ctx->buffer[used], data, size);
                         return;
                 }

                 memcpy(&ctx->buffer[used], data, free);
                 data = (unsigned char *)data + free;
                 size -= free;
                 body(ctx, ctx->buffer, 64);
         }

         if (size >= 64) {
                 data = body(ctx, data, size & ~(unsigned long)0x3f);
                 size &= 0x3f;
         }

         memcpy(ctx->buffer, data, size);
 }

 void MD5_Final(unsigned char *result, MD5_CTX *ctx)
 {
         unsigned long used, free;

         used = ctx->lo & 0x3f;

         ctx->buffer[used++] = 0x80;

         free = 64 - used;

         if (free < 8) {
                 memset(&ctx->buffer[used], 0, free);
                 body(ctx, ctx->buffer, 64);
                 used = 0;
                 free = 64;
         }

         memset(&ctx->buffer[used], 0, free - 8);

         ctx->lo <<= 3;
         ctx->buffer[56] = ctx->lo;
         ctx->buffer[57] = ctx->lo >> 8;
         ctx->buffer[58] = ctx->lo >> 16;
         ctx->buffer[59] = ctx->lo >> 24;
         ctx->buffer[60] = ctx->hi;
         ctx->buffer[61] = ctx->hi >> 8;
         ctx->buffer[62] = ctx->hi >> 16;
         ctx->buffer[63] = ctx->hi >> 24;

         body(ctx, ctx->buffer, 64);

         result[0] = ctx->a;
         result[1] = ctx->a >> 8;
         result[2] = ctx->a >> 16;
         result[3] = ctx->a >> 24;
         result[4] = ctx->b;
         result[5] = ctx->b >> 8;
         result[6] = ctx->b >> 16;
         result[7] = ctx->b >> 24;
         result[8] = ctx->c;
         result[9] = ctx->c >> 8;
         result[10] = ctx->c >> 16;
         result[11] = ctx->c >> 24;
         result[12] = ctx->d;
         result[13] = ctx->d >> 8;
         result[14] = ctx->d >> 16;
         result[15] = ctx->d >> 24;

         memset(ctx, 0, sizeof(MD5_CTX));
 }


utils.h
/* **********************************************************
    UniSync - Universal direcotry sync-diff utility
     http://hyperprog.com

    (C) 2014-2019 Peter Deak (hyper80@gmail.com)

    License: GPLv2  http://www.gnu.org/licenses/gpl-2.0.html
************************************************************* */
#ifndef UNISYNC_UTILS_H
#define UNISYNC_UTILS_H

#include "unisync.h"

char dtoh(int v);
void trimenddir(char *str);
char *unifypath(char *path);
char *chop(char *str);
int my_dtoa(double v,char *buffer,int bufflen,int min,int max,int group);
int gethash(const char *fullpath,char *hexhash,int hashmode=HASH_SHA256,int needprefix = 1);
char read_and_echo_character();

struct PathMakerCacheItem
{
    char path[512];
    struct PathMakerCacheItem* n;
};

class PathMaker
{
public:
    static int mkpath(const char *path,bool contains_filename);
    static void clearCache(void);
private:
    static bool   findPath(char *path);
    static struct PathMakerCacheItem* cache;
};

class FileCopier
{
public:
    static double ckbytes;
    time_t ts,te;

    FileCopier(UniSyncConfig *ucp);
    int copy(const char *source,const char *dest);
    int copy_std(const char *source,const char *dest);
    int copy_spec(const char *source,const char *dest);
    int fixtime(const char *source,const char *dest);
    void resetCounters(void);
    void printStatistics();
    int deletefile(const char *path);
    int deletefolder(const char *path);

private:
    UniSyncConfig *uc;
};

#endif // UNISYNC_UTILS_H


unisync.cpp
/* **********************************************************
    UniSync - Universal direcotry sync-diff utility
     http://hyperprog.com

    (C) 2014-2019 Peter Deak (hyper80@gmail.com)

    License: GPLv2  http://www.gnu.org/licenses/gpl-2.0.html
************************************************************* */
#include <stdio.h>
#include <string.h>
#include <dirent.h>
#include <time.h>
#include <stdlib.h>
#include <ctype.h>
#include <math.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <utime.h>

#include "unisync.h"
#include "utils.h"
#include "catalog.h"

#ifdef _WIN32
#include <windows.h>
#include <locale.h>
#endif

int printhelp(void)
{
    printf("%s - Universal offline/online direcotry sync/diff utility\n\n",PROGRAMNAME);
    printf("  Usage: %s <command> [switches] [other parameters]...\n",PROGRAMCMD);
    printf("    \n");
    printf("Commands:\n");
    printf("  create - Create a catalog file from a directory\n");
    printf("    %s create cat:CATALOGFILE SOURCE_DIRECOTRY [switches]\n",PROGRAMCMD);
    printf("    %s create cat:./mycatalog.usc /STORE/MyPics -md5 -v \n",PROGRAMCMD);
    printf("    %s create cat:./mycatalog.usc \"/Dir with spaces/a\" -md5\n",PROGRAMCMD);
    printf("    \n");
    printf("  diff - Compares two directories and print the differences\n");
    printf("    %s diff SOURCE_DIRECOTRY DESTINATION_DIRECTORY [switches]\n",PROGRAMCMD);
    printf("    %s diff /STORE/MyPics /STORE/BackupMyPics -md5 -vv \n",PROGRAMCMD);
    printf("    \n");
    printf("  catdiff - Compares a directory to catalog\n");
    printf("    %s catdiff cat:CATALOGFILE SOURCE_DIRECOTRY [switches]\n",PROGRAMCMD);
    printf("    %s catdiff cat:./mycatalog.usc /STORE/BackupMyPics -md5 -v \n",PROGRAMCMD);
    printf("    %s catdiff cat:\"./my pics\" /STORE/BackupMyPics -md5 -vv \n",PROGRAMCMD);
    printf("    \n");
    printf("  sync - Syncronize a directory to a directory\n");
    printf("    %s sync SOURCE_DIRECOTRY DESTINATION_DIRECTORY [switches]\n",PROGRAMCMD);
    printf("    %s sync /STORE/MyPics /STORE/BackupMyPics -md5 -vv \n",PROGRAMCMD);
    printf("    %s sync /STORE/MyPics /STORE/BackupMyPics -exclf=Thumbs.db -v\n",PROGRAMCMD);
    printf("    %s sync /STORE/MyPics /STORE/BackupMyPics -exclf=Thumbs.db -i -vv\n",PROGRAMCMD);
    printf("    \n");
    printf("  makeupdate - Create an update package to sync offline directories\n");
    printf("    %s makeupdate cat:CATALOGFILE SOURCE_DIRECOTRY update:UPDATEDIR\n",PROGRAMCMD);
    printf("    %s makeupdate cat:./mycatalog.usc /STORE/MyPics update:/media/pen/upd\n",PROGRAMCMD);
    printf("    \n");
    printf("  makesyncupdate - Create an update package to sync online directories\n");
    printf("    %s makeupdate SOURCE_DIRECOTRY DESTINATION_DIRECTORY update:UPDATEDIR\n",PROGRAMCMD);
    printf("    %s makeupdate /STORE/MyPics /STORE/BackupMyPics update:/media/pen/upd\n",PROGRAMCMD);
    printf("    \n");
    printf("  applyupdate - Apply update package to sync directories\n");
    printf("    %s makeupdate SOURCE_DIRECOTRY update:UPDATEDIR\n",PROGRAMCMD);
    printf("    %s makeupdate /STORE/BackupMyPics update:/media/pen/upd\n",PROGRAMCMD);
    printf("    %s makeupdate /STORE/BackupMyPics update:\"update package\" -vv\n",PROGRAMCMD);
    printf("    \n");
    printf("Switches:\n");
    printf(" -v          - Be verbose\n");
    printf(" -vv         - Be more verbose\n");
    printf(" -md5        - Generate md5 hash to check the file's contents\n");
    printf(" -sha2       - Generate sha256 hash to check the file's contents\n");
    printf(" -nohash     - Don't generate any hash (default)\n");
    printf(" -mtime      - Check/Compare modification times of files\n");
    printf(" -fixtime    - Only in SYNC mode: Fixing file times instead of copy\n");
    printf("               if the files appears to be same according to hashes.\n");
    printf("               (This switch only works with hashes and command=sync)\n");
    printf(" -skiphash   - Don't check hashes even if the catalog contains its.\n");
    printf(" -exclf=EXF  - Exclude file named EXF from every work\n");
    printf(" -excld=EXD  - Exclude directory named EXD from every work\n");
    printf(" -exclp=EXP  - Exclude path matched EXP from every work\n");
    printf(" -std        - Use standard POSIX routines instead of platform dependent codes.\n");
    printf(" -i          - Interactive/paranoid sync mode. Print info and ask before sync.\n");
    printf(" -h          - Print help\n");
    printf(" -version    - Version and author informations\n");
    return 0;
}

int printversionabout(void)
{
    printf("%s - Universal Direcotry Sync-Diff utility\n",PROGRAMNAME);
    printf("Author: Peter Deak (hyper80@gmail.com) http://hyperprog.com/unisync\n");
    printf("Version: %s\n",VERSION);
    printf("License: GPLv2\n");
    return 0;
}

void specify(char *val,const char *name);
void specify_and_canopen(char *val,const char *name);
void dontspecify(char *val,const char *name);

int main(int argi,char **argc)
{
    int r;
    if(argi <= 1)
    {
        printhelp();
        return 0;
    }

    char command[512];
    char sourcedir[512];
    char destdir[512];
    char catalogfile[512];
    char updatedir[512];

    strcpy(command,"");
    strcpy(sourcedir,"");
    strcpy(destdir,"");
    strcpy(catalogfile,"");
    strcpy(updatedir,"");

    int spc = 0;
    char *simpleparams[3];
    simpleparams[0] = command;
    simpleparams[1] = sourcedir;
    simpleparams[2] = destdir;

    UniSyncConfig config;
    int p;
    for(p = 1 ; p < argi ; ++p)
    {
        if(!strcmp(argc[p],"-h") || !strcmp(argc[p],"-help"))
            return printhelp();

        if(!strcmp(argc[p],"-version"))
            return printversionabout();

        if(!strcmp(argc[p],"-v"))
        {
            config.verbose = 1;
            continue;
        }
        if(!strcmp(argc[p],"-vv"))
        {
            config.verbose = 2;
            continue;
        }
        if(!strcmp(argc[p],"-vvv"))
        {
            config.verbose = 3;
            continue;
        }
        if(!strcmp(argc[p],"-md5"))
        {
            config.hashmode = HASH_MD5;
            continue;
        }
        if(!strcmp(argc[p],"-sha2"))
        {
            config.hashmode = HASH_SHA256;
            continue;
        }
        if(!strcmp(argc[p],"-nohash"))
        {
            config.hashmode = HASH_EMPTY;
            continue;
        }
        if(!strcmp(argc[p],"-mtime"))
        {
            config.watchtime = 1;
            continue;
        }
        if(!strcmp(argc[p],"-skiphash"))
        {
            config.skiphash = 1;
            continue;
        }
        if(!strcmp(argc[p],"-fixtime"))
        {
            config.fixmtime = 1;
            continue;
        }
        if(!strcmp(argc[p],"-std"))
        {
            config.usestd = 1;
            continue;
        }
        if(!strcmp(argc[p],"-i"))
        {
            config.interactivesync = 1;
            continue;
        }

        if(!strcmp(argc[p],"-guicall"))
        {
            config.guicall = 1;
            continue;
        }
        if(!strncmp(argc[p],"-exclf",6) || !strncmp(argc[p],"-excld",6) || !strncmp(argc[p],"-exclp",6))
        {
            if(argc[p][6] != '=')
            {
                fprintf(stderr,"Error, Exclude must specified with = sign ( -exclf=file.ext )\n");
                return 1;
            }

            config.exclude = 1;
            ExcludeNames *e=new ExcludeNames();
            if(argc[p][5] == 'f')
                e->typ = EXCL_FILE;
            if(argc[p][5] == 'd')
                e->typ = EXCL_DIR;
            if(argc[p][5] == 'p')
                e->typ = EXCL_PATH;

            strncpy(e->name,argc[p]+7,254);
            if(e->typ == EXCL_PATH)
                unifypath(e->name);

            e->n = NULL;
            if(config.exl == NULL)
            {
                config.exl = e;
            }
            else
            {
                ExcludeNames *r=config.exl;
                while(r->n != NULL)
                    r = r->n;
                r->n = e;
            }
            continue;
        }
        if(!strncmp(argc[p],"-wl",3))
        {
            if(argc[p][3] != '=')
            {
                fprintf(stderr,"Error, Windows locale must specified with = sign ( -wl=locale )\n");
                return 1;
            }

            char slocale[32];
            slocale[0]='.';
            strncpy(slocale+1,argc[p]+4,30);
            #ifdef _WIN32
            if(strcmp(slocale+1,"info")!=0)
            {
                printf("Set locale to: %s\n",slocale+1);
                setlocale(LC_ALL,slocale);
            }
            printf("Locale is: %s\n",setlocale(LC_ALL,NULL));
            #else
            printf("Info: -wl, Windows locale switch is unaffected now.\n");
            #endif
            continue;
        }
        if(!strncmp(argc[p],"cat:",4))
        {
            strncpy(catalogfile,argc[p]+4,510);
            int l=strlen(catalogfile);
            if(l<4 || strcmp(catalogfile+l-4,".usc"))
                strcpy(catalogfile+l,".usc");
            continue;
        }
        if(!strncmp(argc[p],"update:",7))
        {
            strncpy(updatedir,argc[p]+7,510);
            continue;
        }
        if(!strncmp(argc[p],"-",1))
        {
            fprintf(stderr,"Error, unknown switch: \"%s\"\n",argc[p]);
            return 1;
        }
        if(spc < 3)
        {
            strncpy(simpleparams[spc],argc[p],510);
            ++spc;
        }
        else
        {
            fprintf(stderr,"Error, too much parameter passed.\n");
            return 1;
        }
    }

    if(!strcmp(command,"help"))
        return printhelp();
    if(!strcmp(command,"version") || !strcmp(command,"about"))
        return printversionabout();

    trimenddir(sourcedir);
    trimenddir(destdir);
    trimenddir(updatedir);

    if(config.verbose > 1)
    {
        printf("----- PARAMETERS -----\nCommand: %s\nSource: %s\nDestination: %s\nUpdate: %s\nCatalog: %s\n",
                command,sourcedir,destdir,updatedir,catalogfile);
        if(config.exclude)
        {
            ExcludeNames *r=config.exl;
            while(r!= NULL)
            {
                printf("Exclude %s: %s\n",
                        r->typ == EXCL_FILE ? "file" :(r->typ == EXCL_DIR ? "directory" : "path"),
                        r->name);
                r=r->n;
            }
        }
        printf("--- END PARAMETERS ---\n");
        if(config.guicall)
            fflush(stdout);
    }

    // **********************************************************************
    //Switch off fixtime switch in some unwanted situation...
    if(config.fixmtime && strcmp(command,"sync")) // ...when not SYNC command requested
        config.fixmtime = 0;
    if(config.fixmtime && config.skiphash) // ...when skiphash is enabled
        config.fixmtime = 0;
    // **********************************************************************
    if(!strcmp(command,"create"))
    {
        specify(catalogfile,"catalog file");
        specify_and_canopen(sourcedir,"source directory");
        dontspecify(destdir,"directory");
        dontspecify(updatedir,"parameter");

        FILE *catf=NULL;
        catf = fopen(catalogfile,"w");
        if(catf == NULL)
        {
            fprintf(stderr,"Error, Cannot open catalog file for writing: %s\n",catalogfile);
            return 1;
        }

        UniCatalog *catalog = new UniCatalog(&config);
        r = catalog->scandir(sourcedir,catf,false);
        fclose(catf);
        delete catalog;
        return r;
    }
    // **********************************************************************
    if(!strcmp(command,"diff"))
    {
        specify_and_canopen(sourcedir,"source directory");
        specify_and_canopen(destdir,"destination directory");
        dontspecify(catalogfile,"parameter");
        dontspecify(updatedir,"parameter");

        UniCatalog *catalog = new UniCatalog(&config);
        r = catalog->scandir(sourcedir,NULL,true);
        if(r != 0) { delete catalog; return 1; }

        r = catalog->scandir_diff(destdir);
        if(r != 0) { delete catalog; return 1; }

        catalog->diffresultPrint();
        delete catalog;
        return 0;
    }
    // **********************************************************************
    if(!strcmp(command,"catdiff"))
    {
        specify_and_canopen(sourcedir,"source directory");
        specify(catalogfile,"catalog file");
        dontspecify(destdir,"directory");
        dontspecify(updatedir,"parameter");

        UniCatalog *catalog = new UniCatalog(&config);
        r = catalog->read(catalogfile);
        if(r != 0) { delete catalog; return 1; }

        r = catalog->scandir_diff(sourcedir);
        if(r != 0) { delete catalog; return 1; }

        catalog->diffresultPrint();
        delete catalog;
        return 0;
    }
    // **********************************************************************
    if(!strcmp(command,"sync"))
    {
        specify_and_canopen(sourcedir,"source directory");
        specify(destdir,"destination directory");
        dontspecify(catalogfile,"parameter");
        dontspecify(updatedir,"parameter");

        UniCatalog *catalog = new UniCatalog(&config);
        r = catalog->scandir(sourcedir,NULL,true);
        if(r != 0) { delete catalog; return 1; }

        r = catalog->scandir_diff(destdir);
        if(r != 0) { delete catalog; return 1; }

        if(config.interactivesync)
        {
            r = catalog->print_sync_procedures(sourcedir,destdir,DIRECTION_CAT_TO_DIFF);
            char ch;
            printf("Do you really want to start the sync? [y/n]\n");
            ch = read_and_echo_character();
            if(ch != 'y')
            {
                printf("\nSync aborted.\n");
                delete catalog;
                return 0;
            }
        }

        r = catalog->scandir_sync(sourcedir,destdir,DIRECTION_CAT_TO_DIFF);

        delete catalog;
        return r;
    }
    // **********************************************************************
    if(!strcmp(command,"makeupdate"))
    {
        specify_and_canopen(sourcedir,"source directory");
        specify(catalogfile,"catalog file");
        specify(updatedir,"update directory");
        dontspecify(destdir,"directory");

        UniCatalog *catalog = new UniCatalog(&config);
        r = catalog->read(catalogfile);
        if(r != 0) { delete catalog; return 1; }

        r = catalog->scandir_diff(sourcedir);
        if(r != 0) { delete catalog; return 1; }

        r = catalog->make_update_package(sourcedir,updatedir);
        delete catalog;
        return r;
    }
    // **********************************************************************
    if(!strcmp(command,"makesyncupdate"))
    {
        specify_and_canopen(sourcedir,"source directory");
        specify_and_canopen(destdir,"destination directory");
        specify(updatedir,"update directory");
        dontspecify(catalogfile,"parameter");

        UniCatalog *catalog = new UniCatalog(&config);
        r = catalog->scandir(destdir,NULL,true);
        if(r != 0) { delete catalog; return 1; }

        r = catalog->scandir_diff(sourcedir);
        if(r != 0) { delete catalog; return 1; }

        r = catalog->make_update_package(sourcedir,updatedir);
        delete catalog;
        return r;
    }
    // **********************************************************************
    if(!strcmp(command,"applyupdate"))
    {
        specify_and_canopen(sourcedir,"source directory");
        specify_and_canopen(updatedir,"update directory");
        dontspecify(destdir,"directory");
        dontspecify(catalogfile,"parameter");

        UniCatalog *catalog = new UniCatalog(&config);
        r = catalog->apply_update_package(updatedir,sourcedir);
        delete catalog;
        return r;
    }

    fprintf(stderr,"Error, unknown command: %s\n",command);
    return 1;
}

void specify(char *val,const char *name)
{
    if(strlen(val) == 0)
    {
        fprintf(stderr,"Error, You have to specify the %s!\n",name);
        exit(1);
    }
}

void specify_and_canopen(char *val,const char *name)
{
    char val_to_check[512];
    strcpy(val_to_check,val);

#ifdef _WIN32
    //On windows: If the string is a drive letter, for example "D:" append the slash sign to the end
    // (it was removed by trimenddir before)
    if(strlen(val_to_check) == 2 && isalpha(val_to_check[0]) && val_to_check[1] == ':')
    {
        val_to_check[2] = '\\';
        val_to_check[3] = '\0';

    }
#endif

    struct stat s;
    if(strlen(val_to_check) == 0 || stat(val_to_check,&s) != 0)
    {
        fprintf(stderr,"Error, The %s does not specified, not exists or cannot be opened!\n",name);
        exit(1);
    }
}

void dontspecify(char *val,const char *name)
{
    if(strlen(val) > 0)
    {
        fprintf(stderr,"Error, Too much %s specified!\n",name);
        exit(1);
    }
}

UniSyncConfig::UniSyncConfig(void)
{
    guicall = 0;
    verbose = 0;
    watchtime = 0;
    hashmode = HASH_EMPTY;
    skiphash = 0;
    exclude = 0;
    fixmtime = 0;
    usestd = 0;
    interactivesync = 0;
    exl = NULL;
}

void UniSyncConfig::save(void)
{
    s_verbose = verbose;
    s_hashmode = hashmode;
    s_watchtime = watchtime;
    s_skiphash = skiphash;
    s_exclude = exclude;
    s_fixmtime = fixmtime;
}

void UniSyncConfig::restore(void)
{
    verbose = s_verbose;
    hashmode = s_hashmode;
    watchtime = s_watchtime;
    skiphash = s_skiphash;
    exclude = s_exclude;
    fixmtime = s_fixmtime;
}

/* end code */
